modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+
+            methodWriter.mark(nonNull);
+            methodWriter.loadThis();
+            methodWriter.getField(CLASS_TYPE, bindingName, type);
+
+            for (int argument = 0; argument < classBinding.javaMethod.getParameterCount(); ++argument) {
+                visit(irArgumentNodes.get(argument + javaConstructorParameterCount), writeScope);
+            }
+
+            methodWriter.invokeVirtual(type, Method.getMethod(classBinding.javaMethod));
+        } else if (instanceBinding != null) {
+            Type type = Type.getType(instanceBinding.targetInstance.getClass());
+            String bindingName = irInvokeCallMemberNode.getBindingName();
+
+            methodWriter.loadThis();
+            methodWriter.getStatic(CLASS_TYPE, bindingName, type);
+
+            for (int argument = 0; argument < instanceBinding.javaMethod.getParameterCount(); ++argument) {
+                visit(irArgumentNodes.get(argument), writeScope);
+            }
+
+            methodWriter.invokeVirtual(type, Method.getMethod(instanceBinding.javaMethod));
+        } else {
+            throw new IllegalStateException("invalid unbound call");
+        }
+    }
+
+    @Override
+    public void visitFlipArrayIndex(FlipArrayIndexNode irFlipArrayIndexNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+
+        visit(irFlipArrayIndexNode.getChildNode(), writeScope);
+
+        Label noFlip = new Label();
+        methodWriter.dup();
+        methodWriter.ifZCmp(Opcodes.IFGE, noFlip);
+        methodWriter.swap();
+        methodWriter.dupX1();
+        methodWriter.arrayLength();
+        methodWriter.visitInsn(Opcodes.IADD);
+        methodWriter.mark(noFlip);
+    }
+
+    @Override
+    public void visitFlipCollectionIndex(FlipCollectionIndexNode irFlipCollectionIndexNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+
+        visit(irFlipCollectionIndexNode.getChildNode(), writeScope);
+
+        Label noFlip = new Label();
+        methodWriter.dup();
+        methodWriter.ifZCmp(Opcodes.IFGE, noFlip);
+        methodWriter.swap();
+        methodWriter.dupX1();
+        methodWriter.invokeInterface(WriterConstants.COLLECTION_TYPE, WriterConstants.COLLECTION_SIZE);
+        methodWriter.visitInsn(Opcodes.IADD);
+        methodWriter.mark(noFlip);
+    }
+
+    @Override
+    public void visitFlipDefIndex(FlipDefIndexNode irFlipDefIndexNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+
+        methodWriter.dup();
+        visit(irFlipDefIndexNode.getChildNode(), writeScope);
+        Type methodType = Type.getMethodType(
+                MethodWriter.getType(irFlipDefIndexNode.getChildNode().getExpressionType()),
+                MethodWriter.getType(def.class),
+                MethodWriter.getType(irFlipDefIndexNode.getChildNode().getExpressionType()));
```

Comment:
Since `MethodWriter.getType(irFlipDefIndexNode.getChildNode().getExpressionType()))` is repeated, and it's fine to reuse, let's pull it out.

Reply:
Done.